activerecord/lib/active_record/integration.rb

```diff
@@ -151,5 +162,43 @@ def to_param(method_name = nil)
         end
       end
     end
+
+    private
+      # Detects if the value before type cast
+      # can be used to generate a cache_version.
+      #
+      # The fast cache version only works with a
+      # string value directly from the database.
+      #
+      # We also must check if the timestamp format has been changed
+      # or if the timezone is not set to UTC then
+      # we cannot apply our transformations correctly.
+      def can_use_fast_cache_version?(timestamp)
+        timestamp.is_a?(String) &&
+          cache_timestamp_format == :usec &&
+          default_timezone == :utc &&
+          !updated_at_came_from_user?
+      end
+
+      # Converts a raw database string to `:usec`
+      # format.
+      #
+      # Example:
+      #
+      #   timestamp = "2018-10-15 20:02:15.266505"
+      #   raw_timestamp_to_cache_version(timestamp)
+      #   # => "20181015200215266505"
+      #
+      # Postgres truncates trailing zeros, https://bit.ly/2QUlXiZ
+      # to account for this we pad the output with zeros
+      def raw_timestamp_to_cache_version(timestamp)
+        key = timestamp.delete("- :.")
+        padding = 20 - key.length
+        if padding != 0
+          key << "0" * padding
```

Comment:
Can we use `ljust` here - I think it reads better and is slightly faster according to my benchmarks (not that this is a common path to be in).

```suggestion
          key.ljust(20, "0")
```